src/dispatcher.rs

```diff
@@ -0,0 +1,378 @@
+                match response {
+                    Ok(res) => res,
+                    Err(e) => {
+                        tracing::error!(error = %e, "error running WASM module");
+                        // A 500 error makes sense here
+                        let mut srv_err = Response::default();
+                        *srv_err.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+                        srv_err
+                    }
+                }
+        
+            }
+        }
+    }
+}
+
+impl RoutePattern {
+    pub fn parse(path_text: &str) -> Self {
+        match path_text.strip_suffix("/...") {
+            Some(prefix) => Self::Prefix(prefix.to_owned()),
+            None => Self::Exact(path_text.to_owned())
+        }
+    }
+
+    pub fn is_match(&self, uri_fragment: &str) -> bool {
+        match self {
+            Self::Exact(path) => path == uri_fragment,
+            Self::Prefix(prefix) => uri_fragment.starts_with(prefix),
+        }
+    }
+
+    pub fn script_name(&self) -> String {
+        match self {
+            Self::Exact(path) => path.clone(),
+            Self::Prefix(prefix) =>
+                if prefix.starts_with('/') {
+                    prefix.to_owned()
+                } else {
+                    format!("/{}", prefix)
+                }
+        }
+    }
+
+    pub fn original_text(&self) -> String {
+        match self {
+            Self::Exact(path) => path.to_owned(),
+            Self::Prefix(prefix) => format!("{}/...", prefix),
+        }
+    }
+
+    /// Resolve a relative path from the end of the matched path to the end of the string.
+    ///
+    /// For example, if the match is `/foo/...` and the path is `/foo/bar`, it should return `"bar"`,
+    /// but if the match is `/foo/bar` and the path is `/foo/bar`, it should return `""`.
+    pub fn relative_path(&self, uri_path: &str) -> String {
+        let path_base = match self {
+            Self::Exact(path) => path,
+            Self::Prefix(prefix) => prefix,
+        };
+        // It is possible that a root path request matching /... returns a None here,
+        // so in that case the appropriate return is "".
+        uri_path.strip_prefix(path_base).unwrap_or("").to_owned()
+    }
+
+    pub fn append(&self, other: &RoutePattern) -> Self {
+        match self {
+            Self::Exact(path) => other.prepend(path),
+            Self::Prefix(prefix) => other.prepend(prefix),
+        }
+    }
+
+    fn prepend(&self, prefix: &str) -> Self {
+        match self {
+            Self::Exact(subpath) => Self::Exact(format!("{}{}", prefix, subpath)),
+            Self::Prefix(subpath) => Self::Prefix(format!("{}{}", prefix, subpath)),
+        }
+    }
+}
+
+impl RoutingTable {
+    pub fn build(source: &LoadedHandlerConfiguration, global_context: RequestGlobalContext) -> anyhow::Result<RoutingTable> {
```

Comment:
@radu-matei I already chatted with @itowlson about this when we paired to code review, but the pre-caching of modules is now gone(-ish) from the code. We are not sure whether to put that back in, or two switch to the pre-init stuff and plug that in here or at the top `main.rs` level.

Reply:
I take back what I said about the cache. See https://github.com/deislabs/wagi/pull/125#issuecomment-933987027